pkg/operator/agent/agent.go

```diff
@@ -217,40 +221,68 @@ func (a *Agent) discoverFlexvolumeDir() string {
 		return getDefaultFlexvolumeDir()
 	}
 
-	// unmarshall to a NodeConfigKubelet
+	// unmarshal to a NodeConfigKubelet
 	configKubelet := NodeConfigKubelet{}
 	if err := json.Unmarshal(nodeConfig, &configKubelet); err != nil {
 		logger.Warningf("unable to parse node config from Kubelet: %+v", err)
-		return getDefaultFlexvolumeDir()
+	} else {
+		flexvolumeDirPath = configKubelet.ComponentConfig.VolumePluginDir
 	}
-	flexvolumeDirPath = configKubelet.ComponentConfig.VolumePluginDir
-	if flexvolumeDirPath == "" {
-		// unmarshall to a NodeConfigControllerManager
-		configControllerManager := NodeConfigControllerManager{}
-		if err := json.Unmarshal(nodeConfig, &configControllerManager); err != nil {
-			logger.Warningf("unable to parse node config from controller manager: %+v", err)
-			return getDefaultFlexvolumeDir()
-		}
+
+	if flexvolumeDirPath != "" {
+		logFlexvolumeDirSource("NodeConfigKubelet")
+		return flexvolumeDirPath
+	}
+
+	// unmarshal to a NodeConfigControllerManager
+	configControllerManager := NodeConfigControllerManager{}
+	if err := json.Unmarshal(nodeConfig, &configControllerManager); err != nil {
+		logger.Warningf("unable to parse node config from controller manager: %+v", err)
+	} else {
 		flexvolumeDirPath = configControllerManager.ComponentConfig.VolumeConfiguration.FlexVolumePluginDir
-		if flexvolumeDirPath == "" {
-			flexvolumeDirPath = getDefaultFlexvolumeDir()
-		}
 	}
 
-	logger.Infof("using flexvolume dir path: %s", flexvolumeDirPath)
-	return flexvolumeDirPath
+	if flexvolumeDirPath != "" {
+		logFlexvolumeDirSource("NodeConfigControllerManager")
```

Comment:
minor suggestion... You could return two values from the function: `source` and `path`. Then the logging of the source could be done once by the caller instead of everywhere there is a return inside this method.

Reply:
yes, i think i like this approach better to return both the path and the source.